File: java/src/main/java/org/sammancoaching/EnterprisePhonebook.java
package org.sammancoaching;

public class EnterprisePhonebook {
    private final Phonebook phonebook;
    private final Authorizer authorizer;
    private final Alerter alerter;


    public EnterprisePhonebook(Phonebook phonebook, Authorizer authorizer, Alerter alerter) {
        this.phonebook = phonebook;
        this.authorizer = authorizer;
        this.alerter = alerter;
    }

    public String lookup(String name) {
        boolean authorized = authorizer.isAuthorized();
        if (!authorized) {
            throw new RuntimeException("unauthorized lookup");
        }
        return phonebook.lookup(name);
    }



    public void add(String name, String number) {
        var clashes = phonebook.findClashes(number);
        if (clashes.isEmpty()) {
            phonebook.add(name, number);
        } else {
            for (int i = 0; i < clashes.size(); i++) {
                var entry = clashes.get(i);
                var event = new BadPhonebookEntryEvent(phonebook.size(), name, number, entry);
                alerter.sendAlert(event);
            }
        }
    }

    // Getter for testing purposes
    Phonebook getPhonebook() {
        return phonebook;
    }

}

File: kotlin/src/main/kotlin/org/sammancoaching/phonebook/EnterprisePhonebook.kt
package org.sammancoaching.phonebook


class EnterprisePhonebook(
    val phonebook: Phonebook,
    private val authorizer: Authorizer,
    private val alerter: Alerter,
) {

    fun lookup(name: String): String {
        val authorized: Boolean = authorizer.isAuthorized
        if (!authorized) {
            throw RuntimeException("unauthorized lookup")
        }
        return phonebook.lookup(name)
    }

    fun add(name: String, number: String) {
        val clashes = phonebook.findClashes(number)
        if (clashes.isEmpty()) {
            phonebook.add(name, number)
        } else {
            for (i in 0..<clashes.size) {
                val entry = clashes[i]
                val event = BadPhonebookEntryEvent(
                    phonebook.size(),
                    name,
                    number,
                    entry)
                alerter.sendAlert(event)
            }
        }
    }
}